src/openapi-generator.ts

```diff
@@ -729,32 +784,50 @@ export class OpenAPIGenerator {
       const options = this.flattenUnionTypes(zodSchema);
 
       return {
-        anyOf: options.map(schema => this.generateInnerSchema(schema)),
+        anyOf: [
+          ...options.map(schema => this.generateInnerSchema(schema)),
+          ...(isNullable ? [this.mapNullableType(undefined, isNullable)] : []),
+        ],
       };
     }
 
     if (isZodType(zodSchema, 'ZodDiscriminatedUnion')) {
       const options = [...zodSchema.options.values()];
 
       return {
-        oneOf: options.map(schema => this.generateInnerSchema(schema)),
-        discriminator: this.mapDiscriminator(options, zodSchema.discriminator),
+        oneOf: [
+          ...options.map(schema => this.generateInnerSchema(schema)),
+          ...(isNullable ? [this.mapNullableType(undefined, isNullable)] : []),
+        ],
+        discriminator: this.mapDiscriminator(
```

Comment:
Maybe we can move the logic with `isNullable` here instead. I.e `isNullable ? this.mapDiscriminator(...) : undefined`. That way the comment "If schema is nullable we should treat it like a union." would make even more sense here.

Reply:
I think I've made it clearer. I agree it was confusing to have isNullable driving behaviour in different areas.